src/library/scala/concurrent/Future.scala

```diff
@@ -628,17 +700,20 @@ object Future {
    * @param futures   the `IterableOnce` of Futures in which to find the first completed
    * @return          the `Future` holding the result of the future that is first to be completed
    */
-  def firstCompletedOf[T](futures: IterableOnce[Future[T]])(implicit executor: ExecutionContext): Future[T] = {
-    val p = Promise[T]()
-    val firstCompleteHandler = new AtomicReference[Promise[T]](p) with (Try[T] => Unit) {
-      override def apply(v1: Try[T]): Unit = getAndSet(null) match {
-        case null => ()
-        case some => some tryComplete v1
+  final def firstCompletedOf[T](futures: IterableOnce[Future[T]])(implicit executor: ExecutionContext): Future[T] =
+    if (futures.isEmpty) Future.firstCompletedIsEmpty
```

Comment:
I actually think having a stack trace might be helpful for figuring out where you made a bad call to `firstCompletedOf`. I'd love to get others' thoughts on whether they think it's an important cost.

I also notice that its behaviour in the case of an empty `IterableOnce` is not specified; the old one effectively returned `Future.never`, and this one returns a failed `Future` (which I think is better). Do you think it's a good idea to have the behaviour well-specified (rather than just "whatever the implementation does")?

Reply:
@NthPortal I'm definitely OK with throwing a new Exception here as it shouldn't be a hot-path operation. Future.never is also an option… :)

Reply:
I'm gonna ask on Gitter/Discourse to see what people think

Reply:
Sébastien thinks we should preserve 2.12 behaviour, which is (effectively) `Future.never`.

I'm actually thinking now that if it *does* throw, it ought to throw in the calling thread (perhaps using `Predef.require`?), rather than inside the returned `Future`, as it is difficult to distinguish between that failure and one from a `Future` passed in which completed with a failure.

Reply:
Often, when deciding what is the correct behavior for an empty series, there exists a "best answer": use the neutral element of the operation. For example, the n-ary sum ∑ of an empty series is 0, but the n-ary product ∏ of an empty series is 1, which is also "why" a^0 = 1. A neutral element is such that, if you *already* have a series of *n* elements, and you add that *x* to the series, the result of the n-ary operation is unchanged.

For `firstCompletedOf`, what is an element `x` that, no matter what series `xs` you have, `firstCompletedOf(xs)` is always `firstCompletedOf(xs :+ x)`. Well ... it's `Future.never`.

By that logic, the correct answer is actually `Future.never`. QED.

Reply:
Switched to Future.never